src/core/data.rs

```diff
@@ -356,39 +356,40 @@ impl Node {
         format!("{}:{}", self.host.replace(".", "_"), self.port)
     }
 
-    pub(crate) fn set_connection(&self, client: BobClient) {
+    pub(crate) async fn set_connection(&self, client: BobClient) {
         debug!("acquire mutex lock on connection");
-        *self.conn.lock().expect("lock mutex") = Some(client);
+        *self.conn.write().await = Some(client);
     }
 
-    pub(crate) fn clear_connection(&self) {
+    pub(crate) async fn clear_connection(&self) {
         debug!("acquire mutex lock on connection");
-        *self.conn.lock().expect("lock mutex") = None;
+        *self.conn.write().await = None;
     }
 
-    pub(crate) fn get_connection(&self) -> Option<BobClient> {
+    pub(crate) async fn get_connection(&self) -> Option<BobClient> {
         debug!("acquire mutex lock on connection");
-        self.conn.lock().expect("lock mutex").clone()
+        self.conn.read().await.clone()
     }
 
     pub(crate) async fn check(&self, client_fatory: &Factory) -> Result<(), String> {
-        if let Some(conn) = self.get_connection() {
+        if let Some(conn) = self.get_connection().await {
             if let Err(e) = conn.ping().await {
                 debug!("Got broken connection to node {:?}", self);
-                self.clear_connection();
+                self.clear_connection().await;
                 Err(format!("{:?}", e))
             } else {
                 debug!("All good with pinging node {:?}", self);
                 Ok(())
             }
         } else {
             debug!("will connect to {:?}", self);
-            client_fatory
-                .produce(self.clone())
-                .await
-                .map(move |client| {
-                    self.set_connection(client);
-                })
+            let client = client_fatory.produce(self.clone()).await;
```

Comment:
I think, it's better to return error with `?` here. So you won't need `if let` below